### PR TITLE
Recents API Improvements

### LINKED ISSUE #41957: Beef up Recents 🥩
related to https://github.com/metabase/metabase/issues/39528
[product doc ](https://www.notion.so/metabase/Add-Recents-tab-to-the-new-EntityPicker-component-895cd80f2b0f4e56a008a82b436c3a00?pm=c)

For the new recent item picker, we're going to need some changes to the activity/recents api

- We need to track the most recent 20 items that each user has "touched" (ie viewed) of each entity type	
	- question
	- model
	- table (probably don't need right now)
	- dashboard
	- collection
- for a collection to be "recently viewed" [means](https://metaboat.slack.com/archives/C064EB1UE5P/p1714504102513489?thread_ts=1714169076.361039&cid=C064EB1UE5P) that:
	- the collection was created
	- a dashboard or question was saved in the collection
	- a dashboard, question, or collection was moved into the collection
- We also need more information about each item than the current api provides
	- `can_write`
	- `last_touched` or some sort of proxy for sorting by most recently used (i think this is what the timestamp field in the recent_views table is doing)
	- parent information - collection name + authority level (eventually parent db for tables)
- While we're at it, can we make the payload for the recents endpoint conform more closely to search and/or collection items?
	- 	nesting `model_object` inside each recent item is a weird pattern that's used nowhere else and is completely unnecessary, we can just populate a `model` property on each item like we do elsewhere
	- we should probably update the popular items api to match

### Desired RecentItem Shape
(open to discussion)

```ts
interface RecentItem {
  id: number;
  name: string;
  model: string; // should differentiate between regular cards and "dataset" for models
  can_write: boolean;
  display: string; // for questions
  parent_collection: { // for non-tables
    id: number;
    name: string;
    authority_level: "official" | null; // for collections
  } | null;
  database: { // for tables
    id: number;
    name: string; // display name
  } | null;
  authority_level: "official" | null; // for collections
  moderated_status: "verified" | null; // for models
  timestamp: string; // last touched
}
```
```[tasklist]
### Tasks
- [x] update return from /activity/recent_views
- [x] track models as distinct from cards
- [x] update recents to preserve inside category (not just last 30 whatevers, but 20 cards, 20 dashboards, etc...)
- [x] Capture collection uses (add event for collection creation, event for moving card, dash, collection)
- [x] update /popular_items return shape to match recent_views
- [x] https://github.com/metabase/metabase/issues/42665
```

### LINKED ISSUE #42665: Reduce number of queries. We have an N+1 situation

### PR DESCRIPTION
**Pull request by escherize**
Resolves #41957
Fixes https://github.com/metabase/metabase/issues/42665

This addresses an opportunity @dpsutton mentioned about improving endpoint's performance.

Main entrypoint is `recent-views/get-list`.

So, instead of querying for each model one at a time, we get all of the entity info up front for each recent_view in a query for its type. e.g. all `question` and `dataset` ids are selected from `:model/Card`, joined with collection and moderation_review.

This data is indexed by model and id, and post-process uses that data structure to look up relevant values to return the `Item` shape.

- [x] filter all things through `mi/can-read?`
- [x] return _effective-parent_ for collection's parent collection

### CODE REVIEW

File: src/metabase/api/activity.clj [hunk 1]
```diff
@@ -29,9 +29,9 @@
                   :archived
                   :collection.authority_level [:collection.name :collection_name]]
      "table"     [Table
-                  :id :name :db_id
-                  :display_name :initial_sync_status
-                  :visibility_type [:metabase_database.name :database_name]])
+                  :id :name :db_id :active
+                  :display_name [:metabase_database.initial_sync_status :initial-sync-status]
+                  :visibility_type [:metabase_database.name :database-name]])
    (let [model-symb (symbol (str/capitalize model))
```
Comment (escherize) [inline src/metabase/api/activity.clj:35]: The output of this query matches the recent-views shape input, so we don't have to query for any of these details either. (`:active` especailly).

File: src/metabase/models/recent_views.clj [hunk 2]
```diff
@@ -155,80 +162,110 @@
 (defmulti fill-recent-view-info
   "Fills in additional information for a recent view, such as the display name of the object.
 
-  - When called from `GET /popular_items`, the `model_object` field will be present, and should be used instead of
-  querying the database for the object."
+  For most things, we try to gather the information in a single query, but for cards and datasets, but certain things
+  are way easier to check with code (e.g. a collection's parent collection.)"
```
Comment (noahmoss) [inline src/metabase/models/recent_views.clj:166]: I think the wording/grammar here is off?

File: src/metabase/models/recent_views.clj [hunk 3]
```diff
@@ -155,80 +162,110 @@
...
-  (-> (t2/select-one [:model/ModerationReview :status]
-                     {:where [:and
-                              [:= :moderated_item_id model-id]
-                              [:= :moderated_item_type (name model)]
-                              [:= :most_recent true]]})
-      :status))
-
-(defn ellide-archived
-  "Returns nil if the model is archived, otherwise returns the model.
+  ;; user-id is not used, but we need it to memoize the function correctly
```
Comment (noahmoss) [inline src/metabase/models/recent_views.clj:175]: I don't see `user-id` in this function?

File: src/metabase/models/recent_views.clj [hunk 4]
```diff
@@ -155,80 +162,110 @@
...
-    (or (get {"model" :dataset
-              "question" :card} card_type)
+    (or (get {"model" :dataset "question" :card} card_type)
         (keyword model))))
 
 (defmethod fill-recent-view-info :default [m] (throw (ex-info "Unknown model" {:model m})))
 
-(defn get-parent-coll
+(defn- get-parent-coll*
   "Gets parent collection info for a recent view item."
```
Comment (noahmoss) [inline src/metabase/models/recent_views.clj:174]: I feel like this function name and docstring is a little misleading — I was expecting it to take a recent view item and return a collection, but it actually takes a collection ID and fetches its details?

File: src/metabase/models/recent_views.clj [hunk 5]
```diff
@@ -155,80 +162,110 @@
...
-      :status))
-
-(defn ellide-archived
-  "Returns nil if the model is archived, otherwise returns the model.
+  ;; user-id is not used, but we need it to memoize the function correctly
+  [coll-id]
+  (if (nil? coll-id)
+    (root/root-collection-with-ui-details {})
+    (-> (t2/select-one :model/Collection coll-id)
+        (select-keys [:id :name :authority_level])
```
Comment (noahmoss) [inline src/metabase/models/recent_views.clj:180]: can you do this in one call?
```suggestion
    (-> (t2/select-one [:model/Collection :id :name :authority_level] coll-id)
```

File: src/metabase/models/recent_views.clj [hunk 6]
```diff
@@ -155,80 +162,110 @@
...
-
-    ;; The model for this recent view doesn't exist in the database, so we can't check if it's archived, so just
-    ;; filter it out.
-    (not model) nil
-
-    (true? (:archived model)) nil
-
-    (nil? (:archived model))
-    (throw (ex-info "Archived field is nil" {:model model}))))
+  (when (false? (:archived model)) model))
```
Comment (noahmoss) [inline src/metabase/models/recent_views.clj:191]: nit - this feels a little more readable to me using a `when-not`
```suggestion
  (when-not (:archived model) model))
```

File: src/metabase/models/recent_views.clj [hunk 7]
```diff
@@ -155,80 +162,110 @@
...
+                          {:id (:collection-id card)
+                           :name (:collection-name card)
+                           :authority_level (:collection-authority-level card)}
+                          (root/root-collection-with-ui-details {}))}))
 
 (defmethod fill-recent-view-info :dataset [{:keys [_model model_id timestamp model_object]}]
-  (when-let [dataset (ellide-archived
-                      (or model_object (t2/select-one :model/Card model_id)))]
+  (when-let [dataset (ellide-archived (or model_object
+                                          (throw (ex-info "???" {}))
```
Comment (noahmoss) [inline src/metabase/models/recent_views.clj:238]: was this left in accidentally?

File: src/metabase/models/recent_views.clj [hunk 8]
```diff
@@ -298,19 +342,41 @@
                                               [:= :rc.id :rv.model_id]]]
                                  :order-by [[:rv.timestamp :desc]]}))
 
-(defn- post-process [recent-view]
+(mu/defn ^:private model->return-model [model :- :keyword]
+  (if (#{:question} model) :card model))
```
Comment (noahmoss) [inline src/metabase/models/recent_views.clj:346]: nit, but I don't really understand why this pattern would be preferable to a straight `=` when you're only comparing against one value?

File: src/metabase/models/recent_views.clj [hunk 9]
```diff
@@ -298,19 +342,41 @@
...
   [user-id]
-  (into [] (comp (map post-process)
-                 (remove nil?))
-        (do-query user-id)))
+  (let [views (do-query user-id)
+        entity->id->data (get-entity->id->data views)]
+    (->> views
+         (map (partial post-process entity->id->data))
+         (remove nil?)
+         vec)))
```
Comment (noahmoss) [inline src/metabase/models/recent_views.clj:382]: does this need to return a vector?

**Review by noahmoss — APPROVED** [review]

File: src/metabase/models/recent_views.clj (hunk 2 above)
Comment (escherize) [reply src/metabase/models/recent_views.clj:166]: Fixed

File: src/metabase/models/recent_views.clj (hunk 6 above)
Comment (escherize) [reply src/metabase/models/recent_views.clj:191]: but if `(:archived model)` is nil, I want that function to return nil

**Comment by escherize** [comment]
I'm converting this to a draft until I include the improved parent_collection support

File: src/metabase/models/recent_views.clj (hunk 5 above)
Comment (escherize) [reply src/metabase/models/recent_views.clj]: This is true

File: src/metabase/models/recent_views.clj (hunk 7 above)
Comment (escherize) [reply src/metabase/models/recent_views.clj]: yup, I've fixed this since

File: src/metabase/models/recent_views.clj (hunk 8 above)
Comment (escherize) [reply src/metabase/models/recent_views.clj:370]: That's fair, I didn't fix this yet, but I will now.

File: src/metabase/models/recent_views.clj (hunk 9 above)
Comment (escherize) [reply src/metabase/models/recent_views.clj]: nope, fixed!

**Review by noahmoss — APPROVED** [review]

**Pull request merged**